// fl-jsuiauto/cypress/helpers/e2e/token-helper.js

var tokenHelper = function(cy){
    this.getTrustedClientAccessToken = function(clientId, clientSecret){
        var request = {
            'grant_type': 'client_credentials',
            'scope': 'flapi.public flapi.internal',
            'client_id': clientId,
            'client_secret': clientSecret
        };

        return cy.request({
            method: 'POST',
            url: Cypress.env("gatewayUrl") + "/connect/token",
            body: request,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            },
            form: true
        })
            .then((response) => {
                return response.body.access_token;
            });
    };

    this.getResourceOwnerAccessToken = function(username, password, clientId, clientSecret){
        var request = {
            'grant_type': 'password',
            'username': username,
            'password': password,
            'scope': 'flapi.public flapi.internal',
            'client_id': clientId,
            'client_secret': clientSecret
        };

        return cy.request({
            method: 'POST',
            url: Cypress.env("gatewayUrl") + "/connect/token",
            body: request,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            },
            form: true
        })
            .then((response) => {
                return response.body.access_token;
            });
    };
};

module.exports = tokenHelper;
