// fl-jsuiauto/cypress/helpers/e2e/serviceClients/userServiceClient.js

var userServiceClient = function(cy) {
    this.markAccountAsTestAccount = function(flid, accessToken){
        return cy.request({
            method: 'POST',
            url: Cypress.env("gatewayUrl") + '/productaccessservice/v0/api/profiles/testflid',
            body: flid.toString(),
            headers: {
                Authorization: `Bearer ${accessToken}`,
                'Content-Type': 'text/json'
            }
        });
    };

    this.registerUser = function(userToRegister, accessToken){
        var userToReturn;

        return cy.request('POST', Cypress.env("gatewayUrl") + '/api/account/register', userToRegister)
            .then((response) => {
                return response.body.model;
            })
            .then((user) => {
                userToReturn = user;
                return this.markAccountAsTestAccount(cy, user.ProfileId, accessToken);
            })
            .then((markAsTestResponse) => {
                return userToReturn;
            });
    };

    this.addAppUser = function(productId, id, appUser, accessToken){
        return cy.request({
            method: 'PUT',
            url: Cypress.env("gatewayUrl") + `/productaccessservice/v0/api/applicationusers/products/${productId}/users/${id}`,
            body: appUser,
            headers: {
                Authorization: `Bearer ${accessToken}`
            }
        })
            .then((response) => {
                return response.body;
            });
    };

    this.assignFlid = function(type, id, flid, accessToken){
        return cy.request({
            method: 'PUT',
            url: Cypress.env("gatewayUrl") + `/productaccessservice/v0/api/users/${type}/${id}/flid`,
            body: flid.toString(),
            headers: {
                'Authorization': `Bearer ${accessToken}`,
                'Content-Type': 'text/json'
            }
        });
    };

};


module.exports = userServiceClient;
